### PR TITLE
Sunbeam V2 Core Stages

### PR DESCRIPTION
**Pull request by ricd99**
# [FEAT]: Sunbeam V2 Core Stages

- [ ] Module Documentation
- [ ] Method/Class/Function Documentation
- [x] Pull Request Completed 

## What's New
- Seven core Sunbeam stages, from Sunbeam V1, re-implemented to use V2's infrastructure (in stage/v3_0):
array.py
energy_from_integrated_power.py
energy_vol_extrapolated.py
integrated_pack_power.py
latitude_longitude.py
motor_power.py
pack_power.py

- A test suite using pytest for each of the seven re-implemented stages (in tests/v3_0)

## Dependencies
- Several signals were added to the language_localization.py, influxdb_language_localization.toml, and sunbeamdb_language_localization.toml files in the data_tools package (in data_tools/localization). data_tools will need to be updated before these V2 core stages can be used. I'll make a PR in the data_tools repository with the changes.

### CODE REVIEW

File: pipeline/executor.py [hunk 1]
```diff
@@ -24,7 +24,8 @@ def __init__(self, event_name: str, engine: Engine, reprocess: bool = False, deb
         stage_library = StageLibrary(event_manager.get_event_pipeline_edition(event_name))
 
         pipeline_stage_definitions = stage_library.get_stages_by_names(pipeline_stage_names)
-        pipeline_stages = [stage() for stage in pipeline_stage_definitions]
+        kwargs = {} #TODO: store these kwargs (likely going to be hardcoded values) in another file?
```
Comment (jonahjlee) [inline pipeline/executor.py]: Why did you need to add kwargs here?

File: stage/v3_0/array.py
Comment (jonahjlee) [inline stage/v3_0/array.py:1]: Looks great! I like the descriptive names. Not sure if it was like that before, but maybe we can rename frequency to frequency_hz for clarity?

File: pipeline/executor.py (hunk 1 above)
Comment (jonahjlee) [inline pipeline/executor.py]: Why are these kwargs needed?

File: stage/v3_0/array.py [hunk 2]
```diff
@@ -0,0 +1,29 @@
+from data_tools.localization import CanonicalName
+from state.frame import Frame, FrameView
+from typing import ClassVar
+from stage.stage import Stage
+
+class Array(Stage):
+    stage_name: ClassVar[str] = "Array"
+    inputs: ClassVar[list[str]] = [
+        CanonicalName.MPPTInputVoltageA,
```
Comment (jonahjlee) [inline stage/v3_0/array.py:9]: I like the descriptive names!

File: stage/v3_0/integrated_pack_power.py [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+        self.total = 0
+        self.last_timestamp = None
+
+    def run(self, input_frame: FrameView) -> Frame:
+        new_frame = Frame.from_view(input_frame)
+        pack_power = input_frame.read(CanonicalName.PackPower)
+        seconds_per_hour = 3600
+
+        if self.last_timestamp is None:
+            dt = 1.0 / self.frequency # for the first tick, estimate the period as we don't have two timestamps yet
```
Comment (jonahjlee) [inline stage/v3_0/integrated_pack_power.py:23]: Nice! I like the estimation technique.

File: stage/v3_0/integrated_pack_power.py [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+
+class IntegratedPackPower(Stage):
+    stage_name: ClassVar[str] = "IntegratedPackPower"
+    inputs: ClassVar[list[str]] = [CanonicalName.PackPower]
+    outputs: ClassVar[list[str]] = [CanonicalName.IntegratedPackPower]
+    frequency: ClassVar[float] = 5 
+
+    def __init__(self, **kwargs):
+        self.total = 0
+        self.last_timestamp = None
```
Comment (jonahjlee) [inline stage/v3_0/integrated_pack_power.py]: Can we add a type hint for the timestamp?

File: stage/v3_0/integrated_pack_power.py [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+from stage.stage import Stage
+
+class IntegratedPackPower(Stage):
+    stage_name: ClassVar[str] = "IntegratedPackPower"
+    inputs: ClassVar[list[str]] = [CanonicalName.PackPower]
+    outputs: ClassVar[list[str]] = [CanonicalName.IntegratedPackPower]
+    frequency: ClassVar[float] = 5 
+
+    def __init__(self, **kwargs):
+        self.total = 0
```
Comment (jonahjlee) [inline stage/v3_0/integrated_pack_power.py]: Maybe this can be `total_energy` or `total_j`?

File: stage/v3_0/latitude_longitude.py [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+
+
+class LatitudeLongitude(Stage):
+    stage_name: ClassVar[str] = "LatitudeLongitude"
+    inputs: ClassVar[list[str]] = [CanonicalName.LatitudeRaw, CanonicalName.LongitudeRaw]
+    outputs: ClassVar[list[str]] = [CanonicalName.LatitudeFiltered, CanonicalName.LongitudeFiltered]
+    frequency: ClassVar[float] = 3
+
+    def __init__(self, event_name: str | None = None, **kwargs):
+        self.min_lat = self.max_lat = self.min_lon = self.max_lon = None
```
Comment (jonahjlee) [inline stage/v3_0/latitude_longitude.py:20]: wow I didn't know that you could do this in Python

File: stage/v3_0/latitude_longitude.py [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+        
+        coords_path = COORDINATES_DIR / event_name / "coords.toml" # TODO: hardcoded path
+        if not coords_path.exists():
+            return
+        
+        with open(coords_path, "rb") as f:
+            data = tomllib.load(f)
+        
+        # abs because telemetry GPS data removes the sign for lat/lon. we remove sign off track coords to match
+        coords = np.abs(np.array(data["coordinates"]))
```
Comment (jonahjlee) [inline stage/v3_0/latitude_longitude.py:32]: Makes sense. I guess you'd run into issues if you are near zero latitude or zero longitude but that shouldn't be a problem.

File: stage/stage.py [hunk 8]
```diff
@@ -10,7 +10,10 @@ class Stage(ABC):
     stage_name: ClassVar[str]
     inputs: ClassVar[list[CanonicalName]]
     outputs: ClassVar[list[CanonicalName]]
-    rate: ClassVar[float]
+    frequency: ClassVar[float]
```
Comment (jonahjlee) [inline stage/stage.py:13]: Can this be renamed to `frequency_hz`?

File: stage/stage.py [hunk 9]
```diff
@@ -10,7 +10,10 @@ class Stage(ABC):
...
+
+    def __init__(self, **kwargs):
+        pass
```
Comment (jonahjlee) [inline stage/stage.py:16]: What does this do?

File: tests/v3_0/test_energy_from_integrated_power.py
Comment (jonahjlee) [inline tests/v3_0/test_energy_from_integrated_power.py:1]: Really great coverage of different cases here, thanks!

File: tests/v3_0/test_latitude_longitude.py [hunk 10]
```diff
@@ -0,0 +1,252 @@
...
+        frame = make_frame_view({
+            CanonicalName.LatitudeRaw: 49.0,
+            CanonicalName.LongitudeRaw: -123.0,
+        })
+        result = latitude_longitude_stage.run(frame)
+        assert result.read(CanonicalName.LatitudeFiltered) == pytest.approx(49.0)
+        assert result.read(CanonicalName.LongitudeFiltered) == pytest.approx(-123.0)
+
+    def test_in_bounds_no_bounds_configured(self, latitude_longitude_stage):
+        assert latitude_longitude_stage.in_bounds(-999, -999) is True
```
Comment (jonahjlee) [inline tests/v3_0/test_latitude_longitude.py]: You can remove the `is True` here, and for `is False`, instead do `assert not X`

File: tests/v3_0/__init__.py
Comment (jonahjlee) [inline tests/v3_0/__init__.py:1]: Really impressed with the amount of tests you've done to cover different cases!

I especially like the named test which aim to check that the behaviour is correct on edge cases, like `test_ipp_negative_increases_energy`, for example.

Keep in mind that the focus of the tests should be to test that the code behaves correctly - I think some of the tests that check the values of constants are not necessary here (e.g. asserting the number of modules is 32). If we were to change this value to represent a change in the car's design, the code would still be correct but the test would fail.

**Review by jonahjlee — CHANGES_REQUESTED** [review]
Awesome work Ryan! I've added some comments in various files.

Overall:
- I think it looks really clean and I like your implementations for the stages and code style
- Thanks for putting so many tests together! I like the framework that you designed for the tests as well
- There are a few TODOs, left over in the code, are you able to resolve these before merging? It would be nice to combine the config for the car in one place.
- A few more type hints would be helpful, mainly in the stage implementations

File: stage/v3_0/array.py
Comment (ricd99) [reply stage/v3_0/array.py:1]: Makes sense, fixed. I also changed the field name from "frequency" to "frequency_hz" in all of the stages.

File: stage/v3_0/array.py (hunk 2 above)
Comment (ricd99) [reply stage/v3_0/array.py:9]: Thank you!

File: stage/v3_0/latitude_longitude.py (hunk 6 above)
Comment (ricd99) [reply stage/v3_0/latitude_longitude.py:20]: 😄

File: stage/v3_0/latitude_longitude.py (hunk 7 above)
Comment (ricd99) [reply stage/v3_0/latitude_longitude.py:32]: I think I get what you mean by issues near zero lat/lon: if the track goes through both +0.1 and - 0.1 latitude, for example, then the location becomes ambiguous once the sign is stripped?

File: tests/v3_0/test_latitude_longitude.py (hunk 10 above)
Comment (ricd99) [reply tests/v3_0/test_latitude_longitude.py]: Makes sense, fixed

File: stage/stage.py (hunk 8 above)
Comment (ricd99) [reply stage/stage.py:13]: Fixed, all frequency fields renamed to frequency_hz

File: stage/v3_0/integrated_pack_power.py (hunk 3 above)
Comment (ricd99) [reply stage/v3_0/integrated_pack_power.py:23]: Thank you!

File: stage/v3_0/integrated_pack_power.py (hunk 4 above)
Comment (ricd99) [reply stage/v3_0/integrated_pack_power.py]: Fixed. I also added a type hint for self.initial_energy attribute in energy_from_integrated_power because that is the other attribute initially set to None, hence the type is unclear.

File: stage/v3_0/integrated_pack_power.py (hunk 5 above)
Comment (ricd99) [reply stage/v3_0/integrated_pack_power.py]: Makes sense, fixed, I changed it to total_energy.

File: tests/v3_0/test_energy_from_integrated_power.py
Comment (ricd99) [reply tests/v3_0/test_energy_from_integrated_power.py:1]: Thank you!

File: pipeline/executor.py (hunk 1 above)
Comment (ricd99) [reply pipeline/executor.py]: The reason for adding kwargs here, creating a default constructor with kwargs in the Stage class, and adding kwargs to the constructors of the stages I implemented is because of line 27 in the old executor.py: `pipeline_stages = [stage() for stage in pipeline_stage_definitions]`.

This line initializes the stages by calling the constructor with no arguments. However, the LatitudeLongitude stage may need to be called with the `event_name` argument for the coordinate filtering I carried over from the old localization stage. But calling `stage(event_name=...)` here doesn't make sense since the other stages don't need the event_name argument. 

It made the most sense to add a kwargs dictionary and adjust the call in executor.py to be `pipeline_stages = [stage(**kwargs) for stage in pipeline_stage_definitions]`.because I can extract event_name in LatitudeLongitude and ignore it in all the other stages. For the stages that have constructors, I just add **kwargs to the parameters. For stages without constructors, I created a default constructor in the Stage class (that all other stages inherit from) that contains **kwargs in the parameters. The default constructor just passes; its sole purpose is to handle the kwargs right now.

In addition, if future stages need an argument that the other stages don't care about, the other stages don't need to be edited. The only edit necessary is adding the new argument to the kwargs dictionary.

File: stage/v3_0/array.py
Comment (ricd99) [reply stage/v3_0/array.py:1]: I've just realized that renaming frequency to frequency_hz in the stages is probably why I'm seeing errors when I'm trying to run Sunbeam right now. I think the name frequency has to be consistent across several other files I'm not too familiar with (scheduler.py, pipeline.py, protocols.py, ...)? Can @joshuaRiefman confirm?

File: stage/stage.py (hunk 9 above)
Comment (ricd99) [reply stage/stage.py:16]: See my comment in executor.py about kwargs

File: pipeline/executor.py (hunk 1 above)
Comment (ricd99) [reply pipeline/executor.py]: See my comment in the other executor.py thread about kwargs

File: tests/v3_0/__init__.py
Comment (ricd99) [reply tests/v3_0/__init__.py:1]: That makes sense. I've removed those checks as well as some other tests that were checking constants. \

As a side note, for full transparency about the tests, I used AI to help me cover cases I would've missed otherwise. I think AI is pretty good at writing thorough tests, and I'm curious if you or Joshua have thoughts about this?

File: stage/v3_0/array.py
Comment (joshuaRiefman) [reply stage/v3_0/array.py:1]: Yeah, it expects exactly `frequency`. I'm not against changing it to `frequency_hz` but it would need to be changed everywhere and I'm not too sure if there's ambiguity right now (what other units of frequency are there?).

File: tests/v3_0/__init__.py
Comment (joshuaRiefman) [reply tests/v3_0/__init__.py:1]: I actually don't think we necessarily need to worry about checks on the car's design since each test will be testing a specific version of the pipeline (I presume) and so the tests can change with the pipeline as the car changes.

File: stage/v3_0/energy_from_integrated_power.py [hunk 11]
```diff
@@ -0,0 +1,27 @@
...
+
+    def run(self, input_frame: FrameView) -> Frame:
+        new_frame = Frame.from_view(input_frame)
+        ipp = input_frame.read(CanonicalName.IntegratedPackPower)
+        eve = input_frame.read(CanonicalName.EnergyVOLExtrapolated)
+
+        if self.initial_energy is None:
+            self.initial_energy = eve
+        
+        current_energy = self.initial_energy - ipp
```
Comment (joshuaRiefman) [inline stage/v3_0/energy_from_integrated_power.py:23]: This seems to be subtracting a unit of power by a unit of energy, which is not correct. I think you should probably be multiplying the power by the stage's period (`1 / frequency`) to convert the power to energy.

Comment (joshuaRiefman) [reply stage/v3_0/energy_from_integrated_power.py:23]: You could also do something similar to what you do in `IntegratedPackPower` where you explicitly track the timestamps to know how much time has passed between calls.

File: stage/v3_0/latitude_longitude.py (hunk 7 above)
Comment (joshuaRiefman) [reply stage/v3_0/latitude_longitude.py:32]: True, but I don't think we will be near the equator any time soon.

File: pipeline/executor.py (hunk 1 above)
Comment (jonahjlee) [reply pipeline/executor.py]: Thanks for clarifying! I like this as a design pattern.

File: tests/v3_0/__init__.py
Comment (jonahjlee) [reply tests/v3_0/__init__.py:1]: I suppose that makes sense if we are creating a new pipeline for each version of the car, but to me if we have tests that double check these scalar values we are not increasing robustness of the code, and rather just doubling the number of places that need to be updated to make a change.

File: stage/v3_0/energy_from_integrated_power.py (hunk 11 above)
Comment (ricd99) [reply stage/v3_0/energy_from_integrated_power.py:23]: > This seems to be subtracting a unit of power by a unit of energy, which is not correct. I think you should probably be multiplying the power by the stage's period (`1 / frequency`) to convert the power to energy.

I believe ipp is a unit of energy as well. ipp is the output of IntegratedPackPower. In IntegratedPackPower, I multiply power (W) by dt (s) and then converts the seconds to hours to create Wh.  dt is the 1 / frequecy step you mentioned.

Comment (ricd99) [reply stage/v3_0/energy_from_integrated_power.py:23]: > You could also do something similar to what you do in `IntegratedPackPower` where you explicitly track the timestamps to know how much time has passed between calls.

Timestamps were needed to calculate energy in IntegratedPackPower, but timestamps aren't needed for calculations in this stage, so I didn't include them. Unless you suggested timestamps for another reason?
